enhancements/installer/aws-cross-account-dns-zone.md

```diff
@@ -0,0 +1,260 @@
+zone in one account (Account A) while cluster resources are installed in a different
+account (Account B). This enhancement enables OpenShift installation with this cross-account shared-VPC
+architecture by providing a solution to handle authentication to the private-hosted-zone account. 
+
+## Motivation
+
+Organizations utilize AWS shared VPCs to enable some degree of resource sharing with participants while
+maintaining administrative privileges over others. Enabling users to install OpenShift with a private
+hosted zone from a separate account allows DNS resolution for the cluster on the shared network while
+maintaining the least-privilege security principle. An alternative solution would be for the OpenShift
+Installer to create a private hosted zone and attach the zone to the shared VPC. Instead, organizations
+can limit the required permissions and maintain greater administrative control by pre-creating a hosted
+zone and a role with limited permissions in that secondary account.
+
+### User Stories
+
+* As an OpenShift cluster admin, I want to install OpenShift on a shared VPC and utilize a private hosted
+zone that belongs to an account different than the rest of the cluster resources.
+
+### Goals
+
+Utilizing the already existing `platform.aws.hostedZone` field in the install config, users
+should be able to install when passing an ID for a hosted zone that belongs to a different account
+than the primary credentials for installation.
+
+### Non-Goals
+
+Modifying public DNS zones from separate accounts is out of scope.
+
+## Proposal
+
+Cluster admins should create a role in the private hosted zone account (Account A),
+and pass the ARN for that role in the install config. The presence of this roleARN
+indicates to the installer/cluster that the private hosted zone belongs to a different
+account, and that the specified role should be assumed when performing operations on
+the hosted zone.
+
+![alt Architecture diagram](aws_shared_vpc.png)
+
+
+### Workflow Description
+
+**cluster creator** is a human user who has been granted **Account B IAM User**
+
+**shared vpc admin** is a human user with super-user-like privileges across multiple accounts.
+
+1. **shared vpc admin** shares a VPC with Account B.
+2. **shared vpc admin** creates a private hosted zone in Account A and attaches it to the shared VPC.
+3. **shared vpc admin** creates a role in Account A granting **Account B IAM User** route53 and tagging
+permissions in Account A.
+4. **cluster creator** specifies the role from step 3 and subnets from step 1 in the install config.
+5. **cluster creator** installs the cluster.
+
+
+### API Extensions
+
+
+### API: DNS
+
+An `awsPrivateHostedZoneRole` field containing the role ARN would be added to the `DNSSpec` in the [cluster DNS object](https://github.com/openshift/api/blob/894b49f57a15cbce3869961e20cd9d52df6f8b0f/config/v1/types_dns.go#L29).
```

Comment:
I'm not an API designer, but I'd like to generically ask: Would it be better to nest this new platform-specific field under platform-specific object (like `aws`,`gcp`,`azure`...)? So it could be `dns.spec.aws.privateHostedZoneRole` or `dns.spec.providerParameters.aws.privateHostedZoneRole`? Just some examples of this pattern: https://github.com/openshift/api/blob/7313280c2886773591481e6bc5e7b0c9335b8086/operator/v1/types_ingress.go#L452 or https://github.com/openshift/api/blob/7313280c2886773591481e6bc5e7b0c9335b8086/operator/v1/types_csi_cluster_driver.go#L124  or https://github.com/openshift/api/blob/7313280c2886773591481e6bc5e7b0c9335b8086/config/v1/types_ingress.go#L102 or https://github.com/openshift/api/blob/7313280c2886773591481e6bc5e7b0c9335b8086/machine/v1/types_controlplanemachineset.go#L233

If we need to create more of provider-specific-dns configurations, would a nested design be more organized opposed to a flatten configuration structure?

Reply:
I think it makes sense to use a [union](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions) for platform. Thanks for the examples! I will update this.

I was basing this current code on https://github.com/openshift/api/blob/master/operator/v1/types_storage.go#L55, but I prefer Grant's suggestion.

Reply:
@gcs278 I took a stab at updating the enhancement and associated API PR with these suggestions. I went with following the approach of the ingress spec, that you linked.

Reply:
Nice - I looks good to me, feels more organized to me, especially if we need to expand to other platforms.